stdlib/mexpr/ast.mc

```diff
@@ -503,10 +504,10 @@ lang MExprAst =
   CmpIntAst + CmpFloatAst + CharAst + SymbAst + CmpSymbAst + SeqOpAst
 
   -- Patterns
-  + VarPat + SeqTotPat + SeqEdgePat + RecordPat + DataPat + IntPat + CharPat +
+  + NamedPat + SeqTotPat + SeqEdgePat + RecordPat + DataPat + IntPat + CharPat +
   BoolPat + AndPat + OrPat + NotPat
 
   -- Types
-  + FunTypeAst + DynTypeAst + UnitTypeAst + CharTypeAst + StringTypeAst +
+  + FunTypeAst + UnknownTypeAst + UnitTypeAst + CharTypeAst + StringTypeAst +
```

Comment:
Maybe, but isn't it easier to just (implicitly) rely on the typechecker to produce an AST without `Unknown`? I considered doing a similar thing for ANF (create an ANF AST), but ultimately decided against this for simplicity.

Reply:
Whenever we consume a `Type` later on exhaustiveness checking will tell us to consider `Unknown` (once we have an exhaustiveness checker). In my eyes a big reason for creating languages as these compositions is the ability to have slightly different ASTs for different stages, so that the typesystem can know about the absence of the things we have removed in earlier stages, or things we've added in later stages.

It is possible that the way we want to do that is by having some form of subtraction on the composition, as opposed to not adding it, but I'm not entirely sure what that would imply yet.

Reply:
Yes, it's always nice to have static guarantees, as long as the work required for defining the different ASTs is not too cumbersome. I think composition with subtraction could be very useful for this. This is also a bit related to what I wrote in the `mcore` channel on Slack a couple of days ago (overriding included data constructors). Sounds like a prime topic for the meeting!